### PR TITLE
Use _setindex! to fix DimensionMismatch in sparse assignment, Fixes #28963

### LINKED ISSUE #28963: Missing DimensionMismatch for SparseMatrixCSC assignment in 0.7
I've encountered a strange behaviour of assigning to a `SparseMatrixCSC` in 0.7:

```Julia
using SparseArrays
a=spzeros(10,10)
b=sprandn(10,20,0.5)
a[:,:]=b[:,:]
```
This does not give a `DimensionMismatch` as it did in Julia v0.6, and I guess it should definitely give an error.

In 0.7 it just assignes the 1:10,1:10 elements of b to a:
```Julia
b[1:10,1:10] ≈ a #true
```

The broadcasted assignment, however, gives an error:

```Julia
a[:,:].=b[:,:]
ERROR: DimensionMismatch("array could not be broadcast to match destination")
Stacktrace:
 [1] check_broadcast_shape at ./broadcast.jl:452 [inlined]
 [2] check_broadcast_shape at ./broadcast.jl:453 [inlined]
 [3] check_broadcast_axes at ./broadcast.jl:455 [inlined]
 [4] instantiate at ./broadcast.jl:269 [inlined]
 [5] materialize!(::SubArray{Float64,2,SparseMatrixCSC{Float64,Int64},Tuple{Base.Slice{Base.OneTo{Int64}},Base.Slice{Base.OneTo{Int64}}},false}, ::Base.Broadcast.Broadcasted{SparseArrays.HigherOrderFns.SparseMatStyle,Nothing,typeof(identity),Tuple{SparseMatrixCSC{Float64,Int64}}}) at ./broadcast.jl:727
 [6] top-level scope at none:0
```

```
versioninfo()
Julia Version 0.7.0
Commit a4cb80f3ed (2018-08-08 06:46 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 4
```

### PR DESCRIPTION
**Pull request by raghav9-97**
Fixes #28963.

### CODE REVIEW

**Comment by raghav9-97** [comment]
Please review this PR.

**Comment by ViralBShah** [comment]
A lot of people are on vacation this week. Hence reviews are likely to be slow.

Also may I request you to make the PR title and commit have a bit more detail about the issue in addition to the number?

**Comment by ViralBShah** [comment]
`Used _setindex! for checking SparseMatrices compatibility fixing DimensionMismatch in assignment , Fixes #28963`.

There can be many places where dimension mismatch happens. It is important to mention that it is fixing the mismatch in assignment.

**Comment by ViralBShah** [comment]
I've reviewed this and am happy to merge once the commit message and PR titles are updated with a little more info as above.

**Comment by raghav9-97** [comment]
Thanks @ViralBShah for helping out.

**Comment by ViralBShah** [comment]
@fredrikekre are we going to backport to 1.0 as well?

**Comment by fredrikekre** [comment]
Yes.

File: stdlib/SparseArrays/src/sparsematrix.jl [hunk 1]
```diff
@@ -2575,6 +2575,7 @@ function setindex!(A::SparseMatrixCSC{Tv,Ti}, V::AbstractVecOrMat, Ix::Union{Int
     @assert !has_offset_axes(A, V, Ix, Jx)
     (I, J) = Base.ensure_indexable(to_indices(A, (Ix, Jx)))
     checkbounds(A, I, J)
+    Base._setindex!(IndexStyle(A), A, V, to_indices(A, (Ix, Jx))...)
```
Comment (mbauman) [inline stdlib/SparseArrays/src/sparsematrix.jl]: Doesn't this just assign into the array twice?  Once with the generic (slow) dense version, and then once with the optimized sparse version?  We just want to be checking the shape of `V`, no?

Comment (mbauman) [reply stdlib/SparseArrays/src/sparsematrix.jl]: Indeed: before:

```
julia> A = spzeros(1000,1000)
1000×1000 SparseMatrixCSC{Float64,Int64} with 0 stored entries

julia> @btime $A[:,:] = $(spzeros(1000, 1000))
  5.557 μs (5 allocations: 8.39 KiB)
```

After:

```
julia> @btime $A[:,:] = $(spzeros(1000, 1000))
  9.545 ms (5 allocations: 8.39 KiB)
```

Comment (mbauman) [reply stdlib/SparseArrays/src/sparsematrix.jl]: #30923

**Pull request merged**